### PR TITLE
Add a tool to run individual clspv passes.

### LINKED ISSUE #263: Expose an executable capable of running clspv-specific passes
@dnovillo 

Executable like llvm-opt to run individual passes to provide better unit testing.

### PR DESCRIPTION
**Pull request by dnovillo**
This fixes #263.  It adds a wrapper around LLVM's opt that allows
running any of the clspv passes on an LLVM IR module.

In this initial implementation, the developer must:

1- Generate IR using the `-print-before=` or `-print-after=` options to
   clspv.
2- Run clspv-opt on the generated LLVM IR file.

This required some re-factoring of `clspv_core`:

1- All the passes are now in a separate shared library (`clspv_passes`).
   This is required by `opt`.
2- The core library continues to be static.

### CODE REVIEW

**Comment by dnovillo** [comment]
Note.  This is still WIP.  I need to add some tests.  Additionally, I would like to add the capability of starting from a `.cl` file, which means running some of `clspv::Compile`.  But this is minimally functional now.

**Comment by dnovillo** [comment]
> Note. This is still WIP. I need to add some tests. Additionally, I would like to add the capability of starting from a `.cl` file, which means running some of `clspv::Compile`. But this is minimally functional now.

Hm, actually now that I think more about this; perhaps it's best if "starting from a `.cl` file" was always done from `clspv` itself.  This tool will always consume and emit LLVM IR.  At most, I'd give it the option to choose `clspv_passes` and `opt` from different paths, instead of the hardwired values it now has.

Thoughts?

**Comment by alan-baker** [comment]
> Hm, actually now that I think more about this; perhaps it's best if "starting from a .cl file" was always done from clspv itself.

Yes. This tool is intended for debugging and testing. If anything, we could add an option to dump the IR from clspv before any optimizations take place (and terminate?).

**Comment by dnovillo** [comment]
> > Hm, actually now that I think more about this; perhaps it's best if "starting from a .cl file" was always done from clspv itself.
> 
> Yes. This tool is intended for debugging and testing. If anything, we could add an option to dump the IR from clspv before any optimizations take place (and terminate?).

Yeah, I'm thinking of a `--emit-ir` to emit the very first generated IR.  And `--emit-ir-after=`, `--emit-ir-before=` for fine-tuned control.  Would that cover all the cases you are thinking of?

**Comment by alan-baker** [comment]
Yes, I'm also ok with `--emit-ir-after` and `-emit-ir-before` coming in a different commit. I'm ok doing some legwork in the interim to get the IR I'm interested in.

**Comment by dnovillo** [comment]
> Yes, I'm also ok with `--emit-ir-after` and `-emit-ir-before` coming in a different commit. I'm ok doing some legwork in the interim to get the IR I'm interested in.

Yes.  I will do this work in a different commit.  I've added support for command-line arguments to clsvp-opt.  PTAL.

File: tools/clspv-opt/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+add_executable(clspv-opt
+  ${CMAKE_CURRENT_SOURCE_DIR}/main.cpp
+)
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} \
+  -DLLVM_OPT=\\\"${CLSPV_BINARY_DIR}/third_party/llvm/bin/opt\\\" \
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply tools/clspv-opt/CMakeLists.txt]: As part of this, can you a simple lit test that invokes a clspv pass (e.g. spirv producer) so the bots can test this out?

[the remark this answers is not in the document]
Comment (dnovillo) [reply tools/clspv-opt/CMakeLists.txt]: > As part of this, can you a simple lit test that invokes a clspv pass (e.g. spirv producer) so the bots can test this out?

Absolutely.  Added one before I left.  It fails, of course.  I'll update the PR once I have it working.

File: lib/CMakeLists.txt [hunk 2]
```diff
@@ -13,26 +13,28 @@
 # limitations under the License.
 
 add_library(clspv_core
-  ${CMAKE_CURRENT_SOURCE_DIR}/AllocateDescriptorsPass.cpp
   ${CMAKE_CURRENT_SOURCE_DIR}/ArgKind.cpp
-  ${CMAKE_CURRENT_SOURCE_DIR}/ClusterPodKernelArgumentsPass.cpp
-  ${CMAKE_CURRENT_SOURCE_DIR}/ClusterConstants.cpp
   ${CMAKE_CURRENT_SOURCE_DIR}/Compiler.cpp
   ${CMAKE_CURRENT_SOURCE_DIR}/ConstantEmitter.cpp
```
[the remark this answers is not in the document]
Comment (dnovillo) [reply lib/CMakeLists.txt]: > This is used by SPIRVProducerPass. It probably belongs in the passes library.

OK, I'll move it.

[the remark this answers is not in the document]
Comment (dnovillo) [reply lib/CMakeLists.txt]: Actually, it already is in `clspv_passes`.  Did you mean that it should be in `clspv_core`?

**Comment by dnovillo** [comment]
Added some tests and addressed review feedback.  PTAL.

**Comment by dnovillo** [comment]
I've consulted LLVM developers and I've learned that the LLVM libraries are not really setup for being used this way on Windows.  It boils down to: "LLVM would need explicit export annotations to make loadable modules work well."  

There is no support for that, so it's not something we will be able to support easily.  On Mac, I may be able to use the `-undefined dynamic_lookup` flag to the linker, but I am not sure if that's enough.

My plan is, then, to: (1) Do not build `clsvp-opt` on Windows (and possibly Mac), (2) only support it on Linux.

I also need to wait for #305 to land so I can re-test with ToT LLVM.

**Comment by dnovillo** [comment]
@alan-baker, @dneto0  the current version offers support for MacOS and Linux.  Support for Windows requires implementing `opt` inside of `clsvp-opt`.  I will extend `clsvp-opt` to support Windows in a separate PR (#307).

PTAL.

File: lib/CMakeLists.txt [hunk 3]
```diff
@@ -55,42 +70,47 @@ add_library(clspv_core
...
-target_include_directories(clspv_core PRIVATE ${LLVM_INCLUDE_DIRS})
+  target_include_directories(${clspv_lib} PRIVATE ${LLVM_INCLUDE_DIRS})
 
-target_include_directories(clspv_core PRIVATE ${CLANG_INCLUDE_DIRS})
+  target_include_directories(${clspv_lib} PRIVATE ${CLANG_INCLUDE_DIRS})
 
-target_include_directories(clspv_core PUBLIC ${CLSPV_INCLUDE_DIRS})
+  target_include_directories(${clspv_lib} PUBLIC ${CLSPV_INCLUDE_DIRS})
+
+  add_dependencies(${clspv_lib} clspv_baked_opencl_header clspv_c_strings clspv_glsl)
```
Comment (alan-baker) [inline lib/CMakeLists.txt]: I'm not certain these dependencies are required for clspv_passes.

File: tools/clspv-opt/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+if(NOT ENABLE_CLSPV_OPT)
+  return()
+endif()
+
+add_executable(clspv-opt ${CMAKE_CURRENT_SOURCE_DIR}/main.cpp)
+
+if(APPLE)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} \
+    -DLLVM_OPT=\\\"${CLSPV_BINARY_DIR}/third_party/llvm/bin/opt\\\" \
```
Comment (alan-baker) [inline tools/clspv-opt/CMakeLists.txt]: I think this may be incorrect for external llvm builds.

Comment (dnovillo) [reply tools/clspv-opt/CMakeLists.txt]: It was.  Also found that `lib/CMakeLists.txt` had a wrong path to clang when doing out-of-tree builds.  Fixed that too.

File: lib/CMakeLists.txt (hunk 3 above)
Comment (dnovillo) [reply lib/CMakeLists.txt]: Adjusted them.  One is needed by `clspv_core`.  The other two are needed by `clspv_passes`.  Added comments to clarify.

**Comment by dnovillo** [comment]
Ready for another round of reviews.  PTAL.

**Pull request merged**